### ISSUE TITLE
services.msc messed up rendering

### ISSUE THREAD

**Issue by LoganDark** [snowie2000/mactype#915]
https://user-images.githubusercontent.com/4723091/211163067-562f1a21-9820-4da5-81ea-e9789573cc60.mp4

![image](https://user-images.githubusercontent.com/4723091/211163081-5ce1903e-a28b-483b-98cc-04f128de0f6e.png)

**Comment by LoganDark** [comment]
can repro on a completely different computer running Windows 10 2004

![image](https://user-images.githubusercontent.com/4723091/211188628-53c52284-baa5-4b9d-b2a8-c3dd259166fd.png)

**Comment by snowie2000** [comment]
Sounds like you replaced your system fonts?

**Comment by LoganDark** [comment]
> Sounds like you replaced your system fonts?

no I did nothing of the sort. these are 2 different computers, the Windows 11 one is freshly installed

they are both using 4K HiDpi screens at 200% scaling

**Comment by snowie2000** [comment]
So maybe your profile has fontsubtitution enabled. 

The Gui is definitely not rendering with the correct resolution. 

And windows 11 is too buggy to support now.

**Comment by LoganDark** [comment]
> maybe your profile has fontsubtitution enabled

it doesn't. this is a brand new profile. it says `FontSubstitutes=0` in ini

> The Gui is definitely not rendering with the correct resolution.

Turning off mactype renders fine

> And windows 11 is too buggy to support now.

I just posted a screenshot of the same exact issue on windows 10

**Comment by snowie2000** [comment]
Looks like you are using a high dpi display and mmc is not using virtual dpi?

**Comment by LoganDark** [comment]
> Looks like you are using a high dpi display and mmc is not using virtual dpi?

Yes in an earlier comment I said that I am using a 4K HiDPI display

I'm not sure what "virtual dpi" is

adding mmc.exe to unload list allows it to work fine, but it's sad to not have mactype rendering in there. is there any way to make it work?

**Comment by snowie2000** [comment]
Okay, the mmc is virtually scaled up, I can see it. I can't see any reason it was rendered like that. I am also on a high dpi display, though not 4k.

Could you please upload a screenshot services.msc working normally together with your profile?

**Comment by LoganDark** [comment]
> Could you please upload a screenshot services.msc working normally together with your profile?

![image](https://user-images.githubusercontent.com/4723091/211190079-5defa306-1533-4b82-98ac-e62e05a36f72.png)
![image](https://user-images.githubusercontent.com/4723091/211190087-ccb0eadc-3646-421c-ba6d-d5b2ea463f1e.png)

**Comment by LoganDark** [comment]
...oh, here is my profile.

```ini
[General]
HintingMode=1
AntiAliasMode=0
NormalWeight=0
BoldWeight=0
ItalicSlant=0
EnableKerning=1
GammaMode=-1
LcdFilter=0
BolderMode=0
TextTuning=0
TextTuningR=0
TextTuningG=0
TextTuningB=0
GammaValue=1.0
Contrast=1.0
RenderWeight=0.8
Fontlink=1
HookChildProcesses=1
FontLoader=1
FontSubstitutes=0
Shadow=0,0,0,0x0,0,0x0
MaxBitmap=0
DirectWrite=1
HintSmallFont=0
CacheMaxFaces=256
CacheMaxSizes=12554432
CacheMaxBytes=12108864
Name=Default
[UnloadDll]
[exclude]
[FontSubstitutes]
[Individual]
[ExcludeSub]
[DirectWrite]

[Experimental]
ClipBoxFix=1
```

**Comment by snowie2000** [comment]
After carefully inspecting your screenshot and video, I've discovered two unusual things:
1. You are using unsupported registry mode, please don't do that.
2. Your mmc has font scaling enabled.

**Comment by LoganDark** [comment]
> * You are using unsupported registry mode, please don't do that.

I thought registry mode was just manual-only because some novices could mess up the installation. Are you saying it's completely unsupported as in you won't help if that's what I'm using? if I reproduce the issue in some other mode (i.e. MacTray mode) would you help then?

> * Your mmc has font scaling enabled.

Can you PLEASE explain what the heck this means? I have not messed with any of the compatibility or scaling settings with mmc. I JUST have a 4K display, and this exact same issue occurs on 2 completely unrelated machines with completely different 4K displays.

**Comment by snowie2000** [comment]
Let me explain it more clearly.

Mmc is famous for its incompatibility with HiDPI, it's always very blurry in high resolution. However, in your "working" screenshot, all texts are clean and smooth, which only means that a compatible option was set to use "System (enhanced)" scaling. Like this one:
![ss1](https://user-images.githubusercontent.com/11767189/211699717-c9879790-455b-4f8d-8f34-8ff5e1b3eec2.png)

MacType is not fully compatible with enhanced scaling, (but is also not known for causing trouble of your kind), because of that, in the latest several versions, a new mechanism was introduced that all applications with system-enhanced scaling will be ignored automatically. But not in registry mode.

**Comment by LoganDark** [comment]
> However, in your "working" screenshot, all texts are clean and smooth, which only means that a compatible option was set to use "System (enhanced)" scaling

Yes, like I said there is no compatibility setting set at all. In fact, the compatibility settings don't even exist for me.

![image](https://user-images.githubusercontent.com/4723091/211700195-f232cbf2-b05b-461f-b46e-6d8e1fb5a815.png)

There's no way I could have set "System (Enhanced)" because I don't have that option.

This is not something I did. It happens on my other machine which is a completely clean install of Windows 11

**Comment by snowie2000** [comment]
Registry mode is a troublemaker. It works perfectly when a system is configured in a way it likes, and no one knows what kind of system it likes or dislikes😅 because of the unbelievable hardware combination Windows can run and millions of billion of software available on the internet. So trial and error is the only way.

We don't recommend registry mode but since it's the perfect way when it does work, it is still there.

**Comment by LoganDark** [comment]
So the solution is just to keep it disabled for mmc and deal with no mactype there ever? (is there any way to fix this?)

**Comment by snowie2000** [comment]
Hmm, I checked my Windows 10 system and it looks like Microsoft has a special treatment on mmc.

Anyways, just try modes other than registry.

**Comment by LoganDark** [comment]
> just try modes other than registry.

but this will not theme things like explorer and UWP apps, will it? my understanding is that registry is the only mode that works systemwide

**Comment by snowie2000** [comment]
Service mode should work fine with explorer, but nothing works perfect with UWP as I can remember. The process creation with UWP is still a mystery.

**Comment by LoganDark** [comment]
> nothing works perfect with UWP

registry mode does, even with font substitutions

![image](https://user-images.githubusercontent.com/4723091/211701311-509d6bca-c202-4275-af84-7b64abc18f99.png)

![image](https://user-images.githubusercontent.com/4723091/211701372-1e90e729-b09e-4cc6-b069-6a23e38e04fd.png)

**Comment by snowie2000** [comment]
That, is totally unexpected....

**Comment by LoganDark** [comment]
You can see why I'm so hesitant to switch methods then, since this one works perfectly for basically everything except mmc (and some things like window titles, taskbar, certain types of text labels / text boxes)

**Comment by snowie2000** [comment]
So does service mode still mess up your mmc or lags your explorer?

**Comment by LoganDark** [comment]
> So does service mode still mess up your mmc or lags your explorer?

I'll check in a bit

**Comment by snowie2000** [comment]
You see, all modes support UWP apps, it's just that there is currently no known way to hook UWP process creation and run mactype before the actual app code. So it's a game that who runs faster, the mactype core or the font creation in that UWP application, that's why UWP rendering works intermittently in modes other than registry.

**Comment by LoganDark** [comment]
> You see, all modes support UWP apps, it's just that there is currently no known way to hook UWP process creation and run mactype before the actual app code. So it's a game that who runs faster, the mactype core or the font creation in that UWP application, that's why UWP rendering works intermittently in modes other than registry.

uwp apps are started by DcomLaunch service (svchost)

**Comment by snowie2000** [comment]
> 

Wow, how do you know? Is there some sort of docs about that? Cuz when you check the process tree in process-explorer, it's the explorer.exe that created them.

PS: MSEdge and settings are very special UWP apps, they are started by svchost and mactype can hook into them correctly every time.

**Comment by LoganDark** [comment]
> Cuz when you check the process tree in process-explorer, it's the explorer.exe that created them.

wrong

![image](https://user-images.githubusercontent.com/4723091/211703630-7c9fd0c6-19bf-4924-9a32-b8f48ba88abe.png)

![image](https://user-images.githubusercontent.com/4723091/211703668-9083c6d5-6b1c-4860-b942-c610b4d8c4e0.png)

> MSEdge and settings are very special UWP apps

those are not UWP apps, they can run when the UWP runtime is removed

**Comment by snowie2000** [comment]
![wt](https://user-images.githubusercontent.com/11767189/211705067-7b97cb75-ef10-4dea-8b7f-cf1f701d5e30.png)
This is Windows Terminal.

And settings is a UWP app, it's just very special and you can't remove UWP runtime.

**Comment by snowie2000** [comment]
And look likes store apps and stock UWP apps are launched in a different way.

**Comment by LoganDark** [comment]
> ![wt](https://user-images.githubusercontent.com/11767189/211705067-7b97cb75-ef10-4dea-8b7f-cf1f701d5e30.png) This is Windows Terminal.
> 
> And settings is a UWP app, it's just very special and you can't remove UWP runtime.

Windows terminal is not a uwp app either. it uses xaml islands to run uwp controls inside a win32 app

Not everything in WindowsApps is UWP, win32 apps can be installed as appx packages, but appx doesn't necessarily mean UWP

> And look likes store apps and stock UWP apps are launched in a different way.

indeed uwp and appx are completely unlinked. store can install both uwp and non-uwp apps, and uwp apps can be either in an appx package or not in one

**Comment by snowie2000** [comment]
You're a true expert on it!

The thing is that nowadays there are so many "fake UWP" apps in the store, and that's the real problem. Do you know how they are started?

**Comment by LoganDark** [comment]
> The thing is that nowadays there are so many "fake UWP" apps in the store, and that's the real problem. Do you know how they are started?

could you give an example of such app? and can you test if restarting DcomLaunch with mactype service running will fix text rendering in such "fake UWP" apps (or would mactype have to run before services.exe itself in order to hook dcomlaunch?)

My Windows 10 machine does not have a functioning update or microsoft store, but I can test on Windows 11

**Comment by snowie2000** [comment]
Actually, mactype does hook into DCom service and therefore all apps started by it get rendered.

**Comment by seth-js** [comment]
> > nothing works perfect with UWP
> 
> registry mode does, even with font substitutions
> 
> ![image](https://user-images.githubusercontent.com/4723091/211701372-1e90e729-b09e-4cc6-b069-6a23e38e04fd.png)

Sorry for the unrelated reply, but what font is this? It looks really nice.

**Comment by LoganDark** [comment]
> > > nothing works perfect with UWP
> > 
> > 
> > registry mode does, even with font substitutions
> > ![image](https://user-images.githubusercontent.com/4723091/211701372-1e90e729-b09e-4cc6-b069-6a23e38e04fd.png)
> 
> Sorry for the unrelated reply, but what font is this? It looks really nice.

[Inter](https://rsms.me/inter/) with cv11 opentype feature (Mactype does not support opentype features but it should)

**Comment by snowie2000** [comment]
If a feature is supported by freetype, it is supported by Mactype.

But some features like variable font are supported by freetype, but I didn't have enough time fully understand its principle under the hood so...

**Comment by LoganDark** [comment]
> If a feature is supported by freetype, it is not supported by Mactype.

Doesn't mactype use freetype for rendering?

**Comment by snowie2000** [comment]
Oh, sorry, I've edited my reply.
